### PR TITLE
gRPC services to support closing

### PR DESCRIPTION
**Pull request by NiteshKant**
__Motivation__

gRPC implementation does not wire close implementations (by implementing methods in `AsyncCloseable`, `GracefulAutoCloseable` or `AutoCloseable`) defines by services such that server closure will invoke the appropriate close methods on the service. Also, there is no way to specify close semantics while implementing interfaces for each RPC method.

__Modification__

- Defined two new interfaces; `Rpc` and `BlockingRpc` that are appropriately implemented by individual RPC methods.
- Introduced a `wrap()` method for each existing `Route` interface variant such that a lambda implementing the route can also define a detached close implementation and this method will attach the close implementation to the route implementation by wrapping.
- Modified gRPC code generation to utilize the above changes.
- Modified gRPC router to use these new functionality and wire close together through various route layers.
- Added tests to verify each close implementation variant.
- Fixed a bug in blocking -> async service conversions that were not implementing graceful close (required for this change).

__Result__

gRPC services can now implement close methods when required.

### CODE REVIEW

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java [hunk 1]
```diff
@@ -98,30 +108,48 @@ private GrpcRouter(final Map<String, RouteProvider> routes,
...
-            if (request.method() != HttpRequestMethod.POST || (service = allRoutes.get(request.path())) == null) {
-                return notFound.handle(ctx, request, responseFactory);
-            } else {
-                return service.handle(ctx, request, responseFactory);
+        return binder.bindStreaming(new StreamingHttpService() {
+            @Override
+            public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                        final StreamingHttpRequest request,
+                                                        final StreamingHttpResponseFactory responseFactory) {
+                StreamingHttpService service;
```
Comment (Scottmitch) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java]: nit while you are here: make this `final`?

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java [hunk 2]
```diff
@@ -184,24 +212,38 @@ RouteProviders drainRoutes() {
...
+                            try {
+                                final GrpcServiceContext serviceContext =
+                                        new DefaultGrpcServiceContext(request.path(), ctx);
+                                final HttpDeserializer<Req> deserializer =
+                                        serializationProvider.deserializerFor(readGrpcMessageEncoding(request),
+                                                requestClass);
+                                return route.handle(serviceContext, request.payloadBody(deserializer))
+                                        .map(rawResp -> newResponse(responseFactory,
+                                                ctx.executionContext().bufferAllocator())
+                                                .payloadBody(uncheckedCast(rawResp),
```
Comment (Scottmitch) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java]: while you are here: `uncheckedCast` shouldn't be necessary (`Class<Resp>` should be used to give type safety) ... consider removal.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java [hunk 3]
```diff
@@ -214,19 +256,37 @@ RouteProviders drainRoutes() {
...
+                                                                final StreamingHttpResponseFactory responseFactory) {
+                        try {
+                            final GrpcServiceContext serviceContext =
+                                    new DefaultGrpcServiceContext(request.path(), ctx);
+                            final HttpDeserializer<Req> deserializer =
+                                    serializationProvider.deserializerFor(readGrpcMessageEncoding(request),
+                                            requestClass);
+                            final Publisher<Resp> response = route.handle(serviceContext,
+                                    request.payloadBody(deserializer))
+                                    .map(GrpcUtils::uncheckedCast);
```
Comment (Scottmitch) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java]: while you are here: `uncheckedCast` shouldn't be necessary (`Class<Resp>` should be used to give type safety) ... consider removal.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java [hunk 4]
```diff
@@ -214,19 +256,37 @@ RouteProviders drainRoutes() {
...
-                        return succeeded(newErrorResponse(responseFactory, t,
-                                ctx.executionContext().bufferAllocator()));
+                StreamingHttpService service = new StreamingHttpService() {
+                    @Override
+                    public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                                final StreamingHttpRequest request,
+                                                                final StreamingHttpResponseFactory responseFactory) {
+                        try {
+                            final GrpcServiceContext serviceContext =
+                                    new DefaultGrpcServiceContext(request.path(), ctx);
```
Comment (Scottmitch) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java]: future consideration for perf (/cc @jayv) ... `path()` takes some extra parsing and by this point we should have already gone through the router layer (which currently also does `path()`) but I wonder if we could just use the raw request target and avoid the parsing all together (and just bail if there are any additional components of the URL).

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java [hunk 5]
```diff
@@ -292,19 +382,33 @@ public HttpExecutionStrategy serviceInvocationStrategy() {
...
-                            final Resp response = route.handle(serviceContext, request.payloadBody(deserializer));
-                            return newResponse(responseFactory, ctx.executionContext().bufferAllocator())
-                                    .payloadBody(response,
-                                            serializationProvider.serializerFor(serviceContext, responseClass));
-                        } catch (Throwable t) {
-                            return newErrorResponse(responseFactory, t, ctx.executionContext().bufferAllocator());
+                    toStreamingHttpService(new BlockingHttpService() {
+                        @Override
+                        public HttpResponse handle(final HttpServiceContext ctx, final HttpRequest request,
+                                                   final HttpResponseFactory responseFactory) throws Exception {
```
Comment (Scottmitch) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java]: nit: remove `throws Exception`

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java [hunk 6]
```diff
@@ -208,7 +208,7 @@ protected GrpcService newServiceFromRoutes(final AllGrpcRoutes routes) {
             final String path, final GrpcExecutionStrategy executionStrategy,
             final StreamingRoute<Req, Resp> route, final Class<Req> requestClass,
             final Class<Resp> responseClass, final GrpcSerializationProvider serializationProvider) {
-        routeBuilder.addStreamingRoute(path, null, route, requestClass, responseClass,
+        routeBuilder.addStreamingRoute(path, executionStrategy, route, requestClass, responseClass,
```
Comment (Scottmitch) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java]: good catch ... unit test here?

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java [hunk 7]
```diff
@@ -446,6 +446,7 @@ protected GrpcService newServiceFromRoutes(final AllGrpcRoutes routes) {
      * @param <Req> Type of request.
      * @param <Resp> Type of response.
      */
+    @FunctionalInterface
```
Comment (Scottmitch) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java]: @idelpivnitskiy 🎁 👍

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/Rpc.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+import io.servicetalk.concurrent.api.AsyncCloseable;
+import io.servicetalk.concurrent.api.Completable;
+
+import static io.servicetalk.concurrent.api.Completable.completed;
+
+/**
+ * A contract for generated RPC method interface.
+ */
+public interface Rpc extends AsyncCloseable {
```
[the remark this answers is not in the document]
Comment (NiteshKant) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/Rpc.java]: The intent for these new interfaces as well as existing `GrpcService` and `BlockingGrpcService` is to provide a contract for these services such that we can add behaviors in the future (with default implementations) without modifying generated code.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java (hunk 4 above)
Comment (NiteshKant) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java]: We do cache the parsed result for the `URI` inside the request implementation[1] so calling `path()` multiple times would not incur additional parsing. Lets see if this shows up in perf and we can fix, thanks for pointing out 🙏 

[1] https://github.com/apple/servicetalk/blob/master/servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultHttpRequestMetaData.java#L293-L295

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java (hunk 6 above)
Comment (NiteshKant) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRoutes.java]: I have created an issue #907 to add tests to verify the strategy used by routes as @idelpivnitskiy is working on it ATM.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/BlockingRpc.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+package io.servicetalk.grpc.api;
+
+import io.servicetalk.concurrent.GracefulAutoCloseable;
+
+/**
+ * A contract for generated blocking RPC method interface.
+ */
+public interface BlockingRpc extends GracefulAutoCloseable {
```
[the remark this answers is not in the document]
Comment (NiteshKant) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/BlockingRpc.java]: Actually using the `[Blocking]GrpcService` interface here is sufficient, let me do that

File: servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/ClosureTest.java [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+    public static Collection<Boolean> data() {
+        return asList(true, false);
+    }
+
+    @Test
+    public void serviceImplIsClosed() throws Exception {
+        CloseSignal signal = new CloseSignal();
+        TesterService svc = setupCloseMock(mock(TesterService.class), signal);
+        startServerAndClose(new ServiceFactory(svc), signal);
+        verifyClosure(svc, 4 /* 4 rpc methods */);
```
[the remark this answers is not in the document]
Comment (NiteshKant) [reply servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/ClosureTest.java]: It has been our assumption all the way that close methods are idempotent like `AutoCloseable` but ya it is a good idea to be explicit about it in the javadoc. I will do that as a follow up.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/BlockingRpc.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/BlockingRpc.java]: Can we kill these interfaces now?

File: servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/ClosureTest.java [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+    }
+
+    @Parameterized.Parameters(name = "graceful? => {0}")
+    public static Collection<Boolean> data() {
+        return asList(true, false);
+    }
+
+    @Test
+    public void serviceImplIsClosed() throws Exception {
+        CloseSignal signal = new CloseSignal(1);
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/ClosureTest.java]: Should it be 4?

**Comment by idelpivnitskiy** [comment]
@NiteshKant btw, do you want to annotate generated `*Rpc` interfaces with `@FunctionalInterface` as part of this work?

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/BlockingRpc.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (NiteshKant) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/BlockingRpc.java]: Uggh yes ...

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/Rpc.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (NiteshKant) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/Rpc.java]: Update: 

As part of [this dicussion](https://github.com/apple/servicetalk/pull/903#discussion_r362641889) I have removed this new interfaces and using the existing `[Blocking]GrpcService` interfaces instead.

**Comment by NiteshKant** [comment]
>> do you want to annotate generated *Rpc interfaces with @FunctionalInterface as part of this work?

@idelpivnitskiy I would prefer not to increase the scope of this PR now as it is close to merging.

File: servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/ClosureTest.java [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+        CloseSignal signal = new CloseSignal(4);
+        TesterService svc = setupCloseMock(mock(TesterService.class), signal);
+        startServerAndClose(new ServiceFactory(svc), signal);
+        verifyClosure(svc, 4 /* 4 rpc methods */);
+        signal.verifyCloseAtLeastCount(closeGracefully);
+    }
+
+    @Test
+    public void blockingServiceImplIsClosed() throws Exception {
+        CloseSignal signal = new CloseSignal(1);
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/ClosureTest.java]: This also should be `4` :)

Comment (NiteshKant) [reply servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/ClosureTest.java]: Hmm actually I take it back. This and the one in `serviceImplIsClosed()` should be `1` as we only expect 1 close call but due to current implementation detail it is called `4` times (as you correctly pointed out)

**Pull request merged**